### PR TITLE
Allow values to be wrapped prior to serialization in etcd

### PR DESCRIPTION
**Pull request by smarterclayton**
This adds a new value transformer to the etcd2 store that can transform
the value from etcd on read and write. This will allow the store to
implement encryption at rest or otherwise transform the value prior to
persistence.

* [x] etcd3 store
* [x] example of transformation
* [x] partial error handling

This is in support of #12742

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/sig-api-machinery-pr-reviews

File: pkg/storage/etcd/etcd_helper.go [hunk 1]
```diff
@@ -40,15 +40,29 @@ import (
 	"golang.org/x/net/context"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store.
+type ValueTransformer interface {
+	ReadString(string) (string, error)
```
Comment (sttts) [inline pkg/storage/etcd/etcd_helper.go]: Would prefer encode/decode. For Read/Write the signature is unexpected.

File: pkg/storage/etcd/etcd_helper.go [hunk 2]
```diff
@@ -498,6 +528,12 @@ func (h *etcdHelper) GuaranteedUpdate(
 		if err != nil {
 			return err
 		}
+		newBody := string(data)
+
+		body, err := h.transformer.WriteString(newBody)
```
Comment (sttts) [inline pkg/storage/etcd/etcd_helper.go]: A consistent naming of the transformed and raw value would it easier to read. E.g. body = transformed, value = raw, or conversely. Now it's mixed.

File: pkg/storage/etcd/etcd_helper.go [hunk 3]
```diff
@@ -40,15 +40,29 @@ import (
...
+
+func (identityTransformer) ReadString(s string) (string, error)  { return s, nil }
+func (identityTransformer) WriteString(s string) (string, error) { return s, nil }
+
+var IdentityTransformer ValueTransformer = identityTransformer{}
+
 // Creates a new storage interface from the client
 // TODO: deprecate in favor of storage.Config abstraction over time
-func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int) storage.Interface {
+func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quorum bool, cacheSize int, transformer ValueTransformer) storage.Interface {
```
Comment (sttts) [inline pkg/storage/etcd/etcd_helper.go]: Wouldn't it be simpler to decorate `etcdKeysAPI` with the transformation than adding ReadString/WriteString everywhere?

**Comment by timothysc** [comment]
Not really a huge fan of this approach.  It seems like this capability should be able to be built into the client itself.  

/cc @hongchaodeng @xiang90

File: pkg/storage/etcd/etcd_helper.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: This isn't an encoding decoding step - it's a writer / reader wrapper.

File: pkg/storage/etcd/etcd_helper.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: Yeah, I minimized the changes to the code.  I can make the names more explicit.

File: pkg/storage/etcd/etcd_helper.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: It's not going to work for etcd3 - no wrapping is possible today.

**Comment by smarterclayton** [comment]
I explored that but it's not feasible in the etcd3 client as is.  Also, I don't think wrapping an entire client api (including all the issues w.r.t. body compares) is appropriate when the thing being transformed is just a different step.  Also, there's no guarantee a new storage author will be able to wrap their client api.

This originally started in encode/decode, but that doesn't have the right level of visibility to the store for GuaranteedUpdate in order to ensure that idempotence of updates is preserved (we don't want to reuse nonces).

Within the limitations we have, transformation of the body in the store is as close as we get to the actual requirement.

**Comment by smarterclayton** [comment]
Basically, there are three things that look similar but are not:

1. Encoding / decoding wrappers work when the transformation of encoded data is stable - encryption of data at rest is not necessarily stable because we do not want / need to reuse nonces.
2. Wrapping client libraries means you have to wrap the entire surface area of the API and the caller may miss spots - also not all client libraries are wrappable
3. Altering the data at rest *outside* of the semantics of the API requires integration into the storage label.

**Comment by timothysc** [comment]
If the goal is to address #12742, I don't see a reason why the capability couldn't be in the registry get/puts on the secrets resource.

**Comment by liggitt** [comment]
> I don't see a reason why the capability couldn't be in the registry get/puts on the secrets resource.

has to be in the storage layer because of all the plumbing below the generic registry (watch cache, guaranteedupdate, etc)

**Comment by deads2k** [comment]
It seems pretty reasonable to me.  Complicated, potentially one way, but reasonable.  Have you considered what kind of error the API produces if it can't untransform?

File: pkg/storage/etcd/etcd_helper.go [hunk 4]
```diff
@@ -113,7 +128,13 @@ func (h *etcdHelper) Create(ctx context.Context, key string, obj, out runtime.Ob
 		TTL:       time.Duration(ttl) * time.Second,
 		PrevExist: etcd.PrevNoExist,
 	}
-	response, err := h.etcdKeysAPI.Set(ctx, key, string(data), &opts)
+
+	value, err := h.transformer.WriteString(string(data))
+	if err != nil {
+		return err
```
Comment (deads2k) [inline pkg/storage/etcd/etcd_helper.go]: I'd like a typed error.

File: pkg/storage/etcd/etcd_helper.go [hunk 5]
```diff
@@ -283,7 +304,11 @@ func (h *etcdHelper) extractObj(response *etcd.Response, inErr error, objPtr run
 		}
 		return "", nil, fmt.Errorf("unable to locate a value on the response: %#v", response)
 	}
-	body = node.Value
+
+	body, err = h.transformer.ReadString(node.Value)
+	if err != nil {
+		return body, nil, err
```
Comment (deads2k) [inline pkg/storage/etcd/etcd_helper.go]: I think we need a typed error here so that the API can report the issue back to the client which can decide how to proceed.

**Comment by liggitt** [comment]
> Have you considered what kind of error the API produces if it can't untransform?

do we still fail a list if any one item cannot be decoded? this makes it far more likely we'll hit a scenario like that

**Comment by smarterclayton** [comment]
> do we still fail a list if any one item cannot be decoded? 

Yes, we do.  I was debating how we would handle that.  I think we're going to have to do partial decoding and return the right errors on update / create / delete.

File: pkg/storage/etcd/etcd_helper.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: Yeah, was going to be another commit in this PR once I got the basic "ew, gross" reactions out of the way.

File: pkg/storage/etcd/etcd_helper.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: Not sure what you mean by key filtering?  This was not an ACL thing, this was intended to enable encrypt at rest for resources as an optional thing.  Definitely comes with a cost, needed to get to a measurable point.

From a use case point of view the simplest and least impactful is simple encrypt at rest, while others (encrypt before kube, external secret integration) are all reasonable and can be done out of tree, but are more opinionated.  Someone willing to pay the operational cost for encrypt at rest can do so in tree.

File: pkg/storage/etcd/etcd_helper.go (hunk 1 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: I may make this ToStorage and FromStorage to be clearer.

**Comment by smarterclayton** [comment]
@k8s-bot kops aws e2e test this

**Comment by smarterclayton** [comment]
Added example of errors and a test to etcd2 path.  Get, Delete, and Create error.  Delete succeeds (decoding body is not required to delete).  List elides results that have errors and logs them to HandleError, and Watch already does that.  Used the storage.InternalError which seems most closely related to "admin fat fingered the security credentials".

File: pkg/storage/etcd3/store.go [hunk 6]
```diff
@@ -39,15 +39,33 @@ import (
 	"k8s.io/kubernetes/pkg/storage/etcd"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
+	TransformFromStorage([]byte) ([]byte, error)
+	// TransformStringToStorage may transform the provided string into the appropriate form in storage or return an error.
+	TransformToStorage([]byte) ([]byte, error)
```
Comment (sttts) [inline pkg/storage/etcd3/store.go]: This is clearer! 👍

File: pkg/storage/etcd/etcd_watcher.go [hunk 7]
```diff
@@ -71,8 +71,9 @@ type etcdWatcher struct {
 	// Note that versioner is required for etcdWatcher to work correctly.
 	// There is no public constructor of it, so be careful when manipulating
 	// with it manually.
-	versioner storage.Versioner
-	transform TransformFunc
+	versioner        storage.Versioner
+	transform        TransformFunc
```
Comment (sttts) [inline pkg/storage/etcd/etcd_watcher.go]: a short comment wouldn't harm, now that we have two transforming concepts.

File: pkg/storage/etcd/etcd_helper.go [hunk 8]
```diff
@@ -40,15 +41,33 @@ import (
 	etcdutil "k8s.io/kubernetes/pkg/storage/etcd/util"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
+	TransformStringFromStorage(string) (string, error)
```
Comment (sttts) [inline pkg/storage/etcd/etcd_helper.go]: No need for the `String` in the func names.

File: pkg/storage/etcd3/store.go [hunk 9]
```diff
@@ -39,15 +39,33 @@ import (
 	"k8s.io/kubernetes/pkg/storage/etcd"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
```
Comment (sttts) [inline pkg/storage/etcd3/store.go]: a `[]byte` value

File: pkg/storage/etcd3/store.go [hunk 10]
```diff
@@ -39,15 +39,33 @@ import (
...
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
+	// TransformStringFromStorage may transform the provided string from its underlying storage representation or return an error.
```
Comment (sttts) [inline pkg/storage/etcd3/store.go]: same here

File: pkg/storage/etcd/etcd_helper.go (hunk 3 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: K, knowing the motivation now makes more sense.. The original linked issue points out secrets explicitly.  

One thing I'm struggling with in this WIP is why we wouldn't put this behavior into the codec and serialization f(n)s of encode and decode?  As it seems like a natural fit...

File: pkg/storage/etcd/etcd_helper.go (hunk 8 above)
Comment (smarterclayton) [inline pkg/storage/etcd/etcd_helper.go]: I want to ensure that one implementation can satisfy both interfaces (a decrypter) so the names need to be different.

**Comment by sttts** [comment]
I haven't seen an actual encryption transformer implementation yet, but here is a remark:

There is one assumption we cannot make about encryption: it will not return the same encrypted data for the same input. The reason is it will use some random padding behind the actual input to avoid certain attacks. In other words: `encrypt(data)` will change over time. 

This might rule out certain alternative layers to put the encryption in. 

@liggitt can you comment on this, you seem to refer to something similar in https://github.com/kubernetes/kubernetes/pull/40664#issuecomment-276147542

**Comment by smarterclayton** [comment]
Yes, that is why this layer exists.  Because GuaranteedUpdate must be able to compare API bytes -> API bytes for change, and you cannot compare encrypted data like that.

**Comment by smarterclayton** [comment]
This is feature complete (for the intended purpose) - tests and error handlers.

**Comment by smarterclayton** [comment]
@k8s-bot verify test this

**Comment by smarterclayton** [comment]
@k8s-bot you're drunk, go home.

**Comment by smarterclayton** [comment]
Updated with a new boolean (and other fixes on the PR) that allow the Transform layer to signal to GuaranteedUpdate that the data is stale and thus an update *should* be done even if the data hasn't changed.  That would allow keys to be transformed.  This is another reason why Encode wrappers are insufficient and a client wrapper would also be insufficient - GuaranteedUpdate itself is required.

This is complete for the needs of the encrypting store I think.

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go [hunk 11]
```diff
@@ -33,21 +33,42 @@ import (
...
 	"k8s.io/apimachinery/pkg/watch"
 	"k8s.io/apiserver/pkg/storage"
-	utiltrace "k8s.io/apiserver/pkg/util/trace"
 	"k8s.io/apiserver/pkg/storage/etcd"
+	utiltrace "k8s.io/apiserver/pkg/util/trace"
 )
 
+// ValueTransformer allows a string value to be transformed before being read from or written to the underlying store. The methods
+// must be able to undo the transformation caused by the other.
+type ValueTransformer interface {
```
Comment (timothysc) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: Is the only reason this isn't abstracted is because of byte[] vs string?   Can't we de-dupe?

**Review by timothysc — COMMENTED** [review]
Only minor comment.  

@hongchaodeng or @wojtek-t any comments?

File: staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go (hunk 11 above)
Comment (smarterclayton) [inline staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go]: I was hoping we'd drop etcd2 first.  For now, I'd like to avoid unnecessary copies on every store (etcd3 is forcing it), but I would expect that someone would implement both interfaces as efficiently as possible.

**Comment by smarterclayton** [comment]
Any other comments?
